### ISSUE TITLE
[Bug]: redis caching requires async ? 

### ISSUE THREAD

**Issue by liamvdv** [BerriAI/litellm#4511]
### What happened?

Local caching is not really working. E. g. it forces to support both async and sync redis, which is (often) not available.

Issues include 
- forcing some (opaque) async redis, but we only require `completion` / `embeddings` / ... operations which could be checked via `supported_call_types`
- the caching requires litellm[proxy] fields, but it should really be accessible in the normal litellm package

Find a minimal reproducable example below:

```
import litellm
from litellm import completion
from litellm.caching import Cache
from someproject.config import
import time

REDIS_URL = "redis://redis:6379/1"
OPENAI_APIKEY = "..." 

litellm.cache = Cache(type="redis", url=REDIS_URL) # passing `supported_call_types = ["completion"]` has no effect

s = time.time()
x = completion(model="gpt-4o", messages=[{"role": "user", "content": "Hello, how are you? Wink"}], api_key=OPENAI_APIKEY, stream=True)
for m in x:
    print(m)
print(time.time() - s)

s2 = time.time()
x = completion(model="gpt-4o", messages=[{"role": "user", "content": "Hello, how are you? Wink"}], api_key=OPENAI_APIKEY, stream=True)
for m in x:
    print(m)
print(time.time() - s2)
```

Output of pip show litellm
```
Name: litellm
Version: 1.41.3
Location: /Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages
Requires: aiohttp, click, importlib-metadata, jinja2, jsonschema, openai, pydantic, python-dotenv, requests, tiktoken, tokenizers
Required-by:
```

### Relevant log output

```shell
[92m17:18:55 - LiteLLM:ERROR[0m: caching.py:251 - Error connecting to Async Redis client
ERROR:LiteLLM:Error connecting to Async Redis client
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='Hello', role='assistant', function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' just', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' a', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' bundle', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' of', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' code', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933536, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' but', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' here', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' and', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' ready', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' help', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' with', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' anything', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' need', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='.', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' 😊', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' How', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' can', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' assist', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' today', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='?', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZQgqVUPlGUb8g4cxXO1GWTJaf3', choices=[StreamingChoices(finish_reason='stop', index=0, delta=Delta(content=None, role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
1.483821153640747
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='Hello', role='assistant', function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' just', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' lines', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' of', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' code', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933537, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' so', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" don't", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' have', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' feelings', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' but', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' here', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' and', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' ready', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' help', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='.', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' What', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' can', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' assist', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' with', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' today', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='?', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' 😉', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
ModelResponse(id='chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV', choices=[StreamingChoices(finish_reason='stop', index=0, delta=Delta(content=None, role=None, function_call=None, tool_calls=None), logprobs=None)], created=1719933538, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_d576307f90')
0.9468650817871094
[92m17:18:58 - LiteLLM:ERROR[0m: caching.py:440 - LiteLLM Redis Caching: async set() - Got exception from REDIS cannot schedule new futures after interpreter shutdown, Writing value={'timestamp': 1719933538.336788, 'response': '{"id":"chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV","choices":[{"finish_reason":"stop","index":0,"message":{"content":"Hello! I\'m just lines of code, so I don\'t have feelings, but I\'m here and ready to help you. What can I assist you with today? 😉","role":"assistant"}}],"created":1719933537,"model":"gpt-4o","object":"chat.completion","system_fingerprint":"fp_d576307f90","usage":{"completion_tokens":32,"prompt_tokens":15,"total_tokens":47}}'}
ERROR:LiteLLM:LiteLLM Redis Caching: async set() - Got exception from REDIS cannot schedule new futures after interpreter shutdown, Writing value={'timestamp': 1719933538.336788, 'response': '{"id":"chatcmpl-9gZZRHR7BoV3y5SOD11ERS5TfLwSV","choices":[{"finish_reason":"stop","index":0,"message":{"content":"Hello! I\'m just lines of code, so I don\'t have feelings, but I\'m here and ready to help you. What can I assist you with today? 😉","role":"assistant"}}],"created":1719933537,"model":"gpt-4o","object":"chat.completion","system_fingerprint":"fp_d576307f90","usage":{"completion_tokens":32,"prompt_tokens":15,"total_tokens":47}}'}
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-189' coro=<ServiceLogging.async_service_failure_hook() done, defined at /Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/_service_logger.py:89> exception=RuntimeError("can't register atexit after shutdown")>
Traceback (most recent call last):
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/_service_logger.py", line 126, in async_service_failure_hook
    from litellm.proxy.proxy_server import open_telemetry_logger
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/proxy/proxy_server.py", line 57, in <module>
    from apscheduler.schedulers.asyncio import AsyncIOScheduler
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/schedulers/asyncio.py", line 5, in <module>
    from apscheduler.schedulers.base import BaseScheduler
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/schedulers/base.py", line 15, in <module>
    from apscheduler.executors.pool import ThreadPoolExecutor
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/executors/pool.py", line 7, in <module>
    from concurrent.futures.process import BrokenProcessPool
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/process.py", line 101, in <module>
    threading._register_atexit(_python_exit)
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 1504, in _register_atexit
    raise RuntimeError("can't register atexit after shutdown")
RuntimeError: can't register atexit after shutdown
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-192' coro=<ServiceLogging.async_service_success_hook() done, defined at /Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/_service_logger.py:51> exception=RuntimeError("can't register atexit after shutdown")>
Traceback (most recent call last):
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/_service_logger.py", line 79, in async_service_success_hook
    from litellm.proxy.proxy_server import open_telemetry_logger
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/litellm/proxy/proxy_server.py", line 57, in <module>
    from apscheduler.schedulers.asyncio import AsyncIOScheduler
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/schedulers/asyncio.py", line 5, in <module>
    from apscheduler.schedulers.base import BaseScheduler
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/schedulers/base.py", line 15, in <module>
    from apscheduler.executors.pool import ThreadPoolExecutor
  File "/Users/liamvdv/src/github.com/REDACTED/.venv/lib/python3.10/site-packages/apscheduler/executors/pool.py", line 7, in <module>
    from concurrent.futures.process import BrokenProcessPool
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/process.py", line 101, in <module>
    threading._register_atexit(_python_exit)
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 1504, in _register_atexit
    raise RuntimeError("can't register atexit after shutdown")
RuntimeError: can't register atexit after shutdown
```

**Comment by ghost** [comment]
@liamvdv how is this local caching? 

> litellm.cache = Cache(type="redis", url=REDIS_URL)

i can see you've enabled redis here

**Comment by liamvdv** [comment]
@krrishdholakia You refer to 'local' caching in your documentation as caching that is done when python litellm is used directly, i. e. not via http and a proxy endpoint. I've copied that vocabulary.

**Comment by ghost** [comment]
@liamvdv  can you point me to the doc? 


we refer to `local` as in-memory

**Comment by ghost** [comment]
i don't see it here - 
<img width="1091" alt="Screenshot 2024-07-03 at 9 25 52 AM" src="https://github.com/BerriAI/litellm/assets/17561003/206e903f-a1e0-4610-8784-d92aa006b9c1">


https://docs.litellm.ai/docs/caching/all_caches

**Comment by liamvdv** [comment]
https://litellm.vercel.app/docs/caching/local_caching
<img width="964" alt="image" src="https://github.com/BerriAI/litellm/assets/62619917/2e7296f0-cee2-4779-be6e-5791d59e8c88">

**Comment by ghost** [comment]
got it - that document looks outdated and seems like it was unlisted a while back (no navbar) 

I'll clean it up, so it doesn't show up on search and cause confusion

thanks! 


For reference - this is the correct doc https://docs.litellm.ai/docs/caching/all_caches

**Comment by ghost** [comment]
i just ran your snippet FYI

i see a similar log output but it does work for me 

What is the issue that you saw?

**Comment by liamvdv** [comment]
I don't really mind the naming, I just see the issue that using litellm directly (vs. using the http endpoints [proxy]) assumes more that in really requires. I think 
1. the async operations should be shielded and only checked if a* action (e. g. acompletion) is required. 
2.  The service success handler should not always be called, else even for local installations the litellm[proxy] package would be required.

**Comment by ghost** [comment]
<img width="1511" alt="Screenshot 2024-07-03 at 9 28 57 AM" src="https://github.com/BerriAI/litellm/assets/17561003/778cbf99-022d-4d1f-b72f-be065b1c1a51">

**Comment by ghost** [comment]
> I don't really mind the naming, I just see the issue that using litellm directly (vs. using the http endpoints [proxy]) assumes more that in really requires. I think

the async operations should be shielded and only checked if a* action (e. g. acompletion) is required.
The service success handler should not always be called, else even for local installations the litellm[proxy] package would be required.

Got it - working on this now

**Comment by liamvdv** [comment]
try running without optional [proxy] deps.

**Comment by liamvdv** [comment]
@krrishdholakia Thanks for resolving the service hook issue so quickly. **The async-redis issue persists**:

<img width="1034" alt="image" src="https://github.com/BerriAI/litellm/assets/62619917/db7ca783-aa3d-4206-b86e-af7119edfafe">

You still need to add a check that only if a `a[action]` is in the supported operations for caching, async redis is required. 
The error is logged and swallowed, but error reporting tooling will still page.

Please reopen and remove async-redis requirement. 
Best,
Liam

**Comment by ghost** [comment]
able to repro

**Comment by ghost** [comment]
hey @liamvdv we actually need async redis for more than just acompletion 

Example - the router uses it for tracking usage across multiple deployments. 

Pushed a fix to raise debug not error, if it's caused due to no running event loop - https://github.com/BerriAI/litellm/commit/3f83e8a8d420fff4d7f7a635e478034e1cdb0415

**Comment by liamvdv** [comment]
Repro - still not working on newest `litellm==1.41.23`

```
import litellm
from litellm import completion
from litellm.caching import Cache
from botbrains.config import REDIS_URL, OPENAI_APIKEY
import time

# litellm.cache = Cache(type="redis", namespace="litellm", url=REDIS_URL, supported_call_types=["completion", "acompletion"])

litellm.cache = Cache(type="redis", url=REDIS_URL, ttl=2 * 60 * 60)

s = time.time()
x = completion(
    model="gpt-4o",
    messages=[{"role": "user", "content": "Hello, how are you? Wink! I'm at "}],
    api_key=OPENAI_APIKEY,
    stream=True,
)
for m in x:
    print(m)
print(time.time() - s)

s2 = time.time()
x = completion(
    model="gpt-4o",
    messages=[{"role": "user", "content": "Hello, how are you? Wink! a"}],
    api_key=OPENAI_APIKEY,
    stream=True,
)
for m in x:
    print(m)
print(time.time() - s2)
```

```
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='Hello', role='assistant', function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' here', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' and', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' ready', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' help', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='.', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' Sounds', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' like', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219327, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' might', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' have', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' been', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' trying', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' type', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' something', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' else', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='?', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' If', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' need', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' assistance', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' with', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' anything', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' feel', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' free', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' let', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' me', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' know', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
ModelResponse(id='chatcmpl-9ly3y5vnNH6T4cPt8v5O3YxZhgB64', choices=[StreamingChoices(finish_reason='stop', index=0, delta=Delta(content=None, role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219328, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_dd932ca5d1')
2.5690391063690186
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='Hello', role='assistant', function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' just', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' a', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' computer', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' program', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' so', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" don't", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' have', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' feelings', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=',', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' but', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=" I'm", role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' here', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' and', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' ready', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' to', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' help', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' with', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' anything', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' need', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='.', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' How', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' can', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' assist', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' today', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='?', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' 😉', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
ModelResponse(id='chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk', choices=[StreamingChoices(finish_reason='stop', index=0, delta=Delta(content=None, role=None, function_call=None, tool_calls=None), logprobs=None)], created=1721219329, model='gpt-4o', object='chat.completion.chunk', system_fingerprint='fp_5e997b69d8')
1.182621955871582
14:28:49 - LiteLLM:ERROR: caching.py:453 - LiteLLM Redis Caching: async set() - Got exception from REDIS cannot schedule new futures after interpreter shutdown, Writing value={'timestamp': 1721219329.347882, 'response': '{"id":"chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk","choices":[{"finish_reason":"stop","index":0,"message":{"content":"Hello! I\'m just a computer program, so I don\'t have feelings, but I\'m here and ready to help you with anything you need. How can I assist you today? 😉","role":"assistant","tool_calls":null,"function_call":null}}],"created":1721219329,"model":"gpt-4o","object":"chat.completion","system_fingerprint":"fp_5e997b69d8","usage":{"completion_tokens":35,"prompt_tokens":16,"total_tokens":51}}'}
ERROR:LiteLLM:LiteLLM Redis Caching: async set() - Got exception from REDIS cannot schedule new futures after interpreter shutdown, Writing value={'timestamp': 1721219329.347882, 'response': '{"id":"chatcmpl-9ly40NfPg0NvOd0wLE42UIuJzSGmk","choices":[{"finish_reason":"stop","index":0,"message":{"content":"Hello! I\'m just a computer program, so I don\'t have feelings, but I\'m here and ready to help you with anything you need. How can I assist you today? 😉","role":"assistant","tool_calls":null,"function_call":null}}],"created":1721219329,"model":"gpt-4o","object":"chat.completion","system_fingerprint":"fp_5e997b69d8","usage":{"completion_tokens":35,"prompt_tokens":16,"total_tokens":51}}'}
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-213' coro=<ServiceLogging.async_service_failure_hook() done, defined at /Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/litellm/_service_logger.py:89> exception=RuntimeError("can't register atexit after shutdown")>
Traceback (most recent call last):
  File "/Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/litellm/_service_logger.py", line 126, in async_service_failure_hook
    from litellm.proxy.proxy_server import open_telemetry_logger
  File "/Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/litellm/proxy/proxy_server.py", line 52, in <module>
    from apscheduler.schedulers.asyncio import AsyncIOScheduler
  File "/Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/apscheduler/schedulers/asyncio.py", line 5, in <module>
    from apscheduler.schedulers.base import BaseScheduler
  File "/Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/apscheduler/schedulers/base.py", line 15, in <module>
    from apscheduler.executors.pool import ThreadPoolExecutor
  File "/Users/liamvdv/src/github.com/r/.venv/lib/python3.10/site-packages/apscheduler/executors/pool.py", line 7, in <module>
    from concurrent.futures.process import BrokenProcessPool
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/process.py", line 101, in <module>
    threading._register_atexit(_python_exit)
  File "/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 1504, in _register_atexit
    raise RuntimeError("can't register atexit after shutdown")
RuntimeError: can't register atexit after shutdown
```

**Comment by ghost** [comment]
Hey @liamvdv i just ran your script in a colab env, and i can't repro the error. (screenshot attached). 

Code: https://colab.research.google.com/drive/1-oAWLpgtLL-C0Wn7CvZNah4f1gbBQ-VH?usp=sharing

<img width="1450" alt="Screenshot 2024-07-17 at 10 42 25 AM" src="https://github.com/user-attachments/assets/7d656930-f6cb-4ef3-99ba-50803c59c42c">

Are you sure you're on a later version of litellm? 

If so, can you share a colab with consistent repro?

**Comment by liamvdv** [comment]
Colab are the wrong environment here, as the kernel (inc. interpreter) keeps running. The issue appears to happen on immediate shutdowns.
Try in a non-ipy environment.

**Comment by ghost** [comment]
got it - i'll try local python env

**Comment by liamvdv** [comment]
And yes, I'm on the latest version ✅

**Comment by ghost** [comment]
i don't see this on local either 
<img width="1619" alt="Screenshot 2024-07-17 at 10 49 06 AM" src="https://github.com/user-attachments/assets/36430cf9-0cca-4edd-9c0f-475b6b9d05ed">

**Comment by ghost** [comment]
nvm i have repro

**Comment by jscheel** [comment]
I'm still seeing this on 1.44.22

**Comment by ghost** [comment]
hi @jscheel can you create a new issue with logs + steps for repro
